programs/ziti-edge-tunnel/log_utils.c

```diff
@@ -0,0 +1,308 @@
+
+    uv_timeval64_t file_time;
+    uv_gettimeofday(&file_time);
+    start_time = calloc(1, sizeof(struct tm));
+#if _WIN32
+    _gmtime32_s(start_time, &file_time.tv_sec);
+#else
+    gmtime_r(&file_time.tv_sec, start_time);
+#endif
+
+    delete_older_logs(ziti_loop);
+    multi_writer = is_multi_writer;
+
+    log_flusher = calloc(1, sizeof(uv_check_t));
+    uv_check_init(ziti_loop, log_flusher);
+    log_flusher->data = start_time;
+    uv_unref((uv_handle_t *) log_flusher);
+    uv_check_start(log_flusher, flush_log);
+
+
+    log_filename = get_log_filename();
+
+    if (!open_log(log_filename)) {
+        return false;
+    }
+#if _WIN32
+    SvcReportEvent(TEXT(log_filename), EVENTLOG_INFORMATION_TYPE);
+#endif
+    return true;
+}
+
+struct tm* get_log_start_time() {
+    return start_time;
+}
+
+char* get_log_file_name(){
+    return log_filename;
+}
+
+static const char* parse_level(int level) {
+    const char* err_level;
+    switch(level) {
+        case 0:
+            err_level = "FATAL";
+            break;
+        case 1:
+            err_level = "ERROR";
+            break;
+        case 2:
+            err_level = "WARN";
+            break;
+        case 3:
+            err_level = "INFO";
+            break;
+        case 4:
+            err_level = "DEBUG";
+            break;
+        case 5:
+            err_level = "VERBOSE";
+            break;
+        case 6:
+            err_level = "TRACE";
+            break;
+        default:
+            err_level = "UNKNOWN";
+    }
+    return err_level;
+}
+
+void ziti_log_writer(int level, const char *loc, const char *msg, size_t msglen) {
+    char curr_time[32];
+    uv_timeval64_t now;
+    uv_gettimeofday(&now);
+    struct tm *tm = gmtime(&now.tv_sec);
+
+    snprintf(curr_time, sizeof(curr_time), "%04d-%02d-%02dT%02d:%02d:%02d.%03dZ",
+             1900 + tm->tm_year, tm->tm_mon + 1, tm->tm_mday,
+             tm->tm_hour, tm->tm_min, tm->tm_sec, now.tv_usec / 1000
+    );
+
+    if ( ziti_tunneler_log != NULL) {
+        fprintf(ziti_tunneler_log, "\n[%s] %7s %s ", curr_time, parse_level(level), loc);
+        fwrite(msg, 1, msglen, ziti_tunneler_log);
+    }
+
+    if(multi_writer) {
+        printf("\n[%s] %7s %s %.*s", curr_time, parse_level(level), loc, msglen, msg);
+    }
+
+}
+
+bool open_log(char* log_filename) {
+    if((ziti_tunneler_log=fopen(log_filename,"a")) == NULL) {
+        printf("Could not open logs file %s, due to %s", log_filename, strerror(errno));
+        return false;
+    }
+    return true;
+}
+
+void close_log() {
+    if (ziti_tunneler_log != NULL) {
+        fclose(ziti_tunneler_log);
+        ziti_tunneler_log = NULL;
+    }
+    if (log_filename != NULL) {
+        free(log_filename);
+        log_filename = NULL;
+    }
+}
+
+void stop_log_check() {
+    if (log_flusher != NULL) {
+        uv_check_stop(log_flusher);
+        free(log_flusher->data);
+        free(log_flusher);
+    }
+}
+
+void rotate_log() {
```

Comment:
this function does not seem to do what is supposed to:
- close_log() closes log file and frees log_filename
- open_log() would fail because get_log_filename() would return NULL